roachtest: add post-process step for perf metrics

sambhav-jain-16 opened the pull request:
There are many tests in which had some custom post processing done in
`roachperf` to emit the metrics.

To add this facility  in the roachtests itself, this change aims to :

1. Adds 2 new user facing structs `HistogramMetric`, that gives the user
access to the histogram metrics generated by the workload and
`AggregatedPerfMetrics` which user can return with their aggregated metric.
2. Introduce a new `PostProcessPerfMetrics func(testName string, metric *roachtestutil.HistogramMetric) (roachtestutil.AggregatedPerfMetrics, error)` step in test spec where the test writer can do custom
aggregation and provide the aggregated metric. If the test has not
implemented this step, the runner will use the `DefaultPostProcess`
function.
3. The step creates a new openmetrics file with the name
`aggregated_stats.om` that will have the aggregated metrics.
4. Implement `PostProcessPerfMetrics` in some roachtests that do
post processing in roachperf. The implementation is borrowed from `roachperf` for each test.

There are some `roachtests` for which `PostProcessPerfMetrics` is not implemented. It will be done in subsequent PRs

Epic: https://cockroachlabs.atlassian.net/browse/CRDB-41852

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/138617)
<!-- Reviewable:end -->

sambhav-jain-16 committed:
roachtest: add post-process step for perf metrics

There are many tests in which had some custom post processing done in
`roachperf` to emit the metrics.

To add this facility  in the roachtests itself, this change aims to :

1. Adds 2 new user facing structs `HistogramMetric`, that gives the user
access to the histogram metrics generated by the workload and
`AggregatedPerfMetrics` which user can return with their aggregated
metric.
2. Introduce a new `PostProcessPerfMetrics func(testName string, metric
*roachtestutil.HistogramMetric) (roachtestutil.AggregatedPerfMetrics,
error)` step in test spec where the test writer can do custom
aggregation and provide the aggregated metric. If the test has not
implemented this step, the runner will use the `DefaultPostProcess`
function.
3. The step creates a new openmetrics file with the name
`aggregated_stats.om` that will have the aggregated metrics.
4. Implement `PostProcessPerfMetrics` in some roachtests that
do post processing in roachperf. The implementation is borrowed from
`roachperf` for each test.

There are some `roachtests` for which `PostProcessPerfMetrics` is not
implemented. It will be done in subsequent PRs

Epic: https://cockroachlabs.atlassian.net/browse/CRDB-41852

Release note: None

sambhav-jain-16 commented:
Openmetrics run: https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/18851548?hideProblemsFromDependencies=false&hideTestsFromDependencies=false

sambhav-jain-16 commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -1409,6 +1412,13 @@ func (r *testRunner) runTest(
 		}
+
+		if t.spec.Benchmark && t.ExportOpenmetrics() {
+			if err := r.postProcessPerfMetrics(ctx, t, c); err != nil {
```
That's a good point. 
Let me take a look and refactor my changes into the `getPerfArtifacts` function you mentioned.
Thanks for pointing out !

sambhav-jain-16 commented on pkg/cmd/roachtest/test_runner.go:
```diff
@@ -1984,6 +2005,114 @@ func (r *testRunner) maybeInitSideEyeClient(ctx context.Context, l *logger.Logge
 	return token
 }
 
+func (r *testRunner) postProcessPerfMetrics(
```
Yes, it will be called for every test and only one file will be created that will have all the aggregated metrics

sambhav-jain-16 commented on pkg/cmd/roachtest/tests/tpchbench.go:
```diff
@@ -169,6 +171,42 @@ func registerTPCHBenchSpec(r registry.Registry, b tpchBenchSpec) {
+			finalP50Value := 0.0
+			for _, summary := range histograms.Summaries {
+
+				// Now since we have multiple p50(median) latencies for the run, we want to calculate a cumulative median value
```
Actually I just checked `roachperf` aggregation for this again. Again looks like it just outputs the p50 latency for each benchmark.
The actual aggregated value is the mean of all latencies. So I think I'll remove this part of code so.

sambhav-jain-16 commented:
TFTR!

bors r=@csgourav
